teps/0154-concise-remote-resolver-syntax.md

```diff
@@ -0,0 +1,340 @@
+
+## Motivation
+
+The current remote resolution syntax is quite verbose. Previously there used to be a one-liner that users could provide to the resolver when referencing a `Task` or a `Pipeline`. Reducing the current syntax (which could get to multiple lines very easily) down to a one-liner would help reduce the already quite verbose Tekton Spec.
+
+### Goals
+
+* Provide a way to pass the one-liner resolver-reference to the resolvers.
+* Provide guidelines on how resolvers can format the expected resolver-reference.
+* Define the formats for the existing Tekton's resolvers.
+
+### Non-Goals
+
+### Use Cases
+
+* As a task/pipeline author, I would like to reduce the verbosity of the yaml.
+
+### Requirements
+
+* The new resolver-reference should not be backwards incompatible. It should be an additive change.
+* The new resolver-reference should be optional. 
+
+## Proposal
+
+### Extending the `resolver` field
+
+Current Remote Resolution syntax looks like follows:
+
+```yaml
+taskRef:
+  resolver: <resolver-name>
+  params:
+    - name: ...
+      value: ...
+    ...
+``` 
+
+We propose adding a new optional field `url` whose value is [resolver-reference](resolver-reference). This field can be used instead of (or in addition to) the `params`:
+
+```yaml
+taskRef:
+  resolver: <resolver-name>
+  url: <resolver-reference>
+```
+
+In turn, the Remote Resolver will parse the `url` and extract the necessary information it needs to perform resolution.
+ 
+### Resolver Reference
+
+[Package-URL](https://github.com/package-url/purl-spec) is a format that we have begun to use in Tekton Artifacts. To enable a consistent concise format for across resolvers, we propose taking purl as an inspiration to define our format. The reason for not using purl directly is that it is too rigid for the different types of resolvers.
+
+To provide a consistent format across resolvers, we suggest the following `remote-resolver` template: `<scheme>://<location>[@<version>][#<selector>][?<params>]`.
+
+Where, `scheme` is required while `location`, `version`, `selector` and `params` are all optional and should be parsed by the `Remote Resolvers` according to their needs. Additionally, the value of the `url` must be `url encoded`.
+
+The `Remote Resolver` must define the format using the above `remote-resolver` template. The `Remote Resolver` is be responsible for parsing the `resolver-reference` and extracting the necessary information. The Tekton controller will perform no validation on the value of the `url`. It will simply pass it as is in the `Remote Resolution CRD`. 
+
+Examples:
+```
+https://github.com/wlynch/test@main#task.yaml
+git://github.com/wlynch/test@main#task.yaml
+oci://wlynch/image:v1@sha256:...#task/foo?secret=my-secret
+https://example.com/task.yaml
+http://10.5.3.2:8080/task.yaml
+artifact/tasks-catalog/tasks/task@v1.0 # e.g. for the hub resolver where there is no need for a scheme
+```
+
+#### Git Resolver
```

Comment:
I think that's up to the resolver to define. So yes, we could choose to add that here. The question is how would the resolver tell the difference between the different cases.

Reply:
that's where the "problem" may lie. In my head `github://…` is a different resolver than `git+https` or `https`, …
I commented a bit about that in https://github.com/tektoncd/pipeline/issues/7508 😬

Reply:
To me `github://`, `git+https://`, `https` are schemes. I don't think a single resolver has to only be able to handle one scheme. For example, the git resolver could handle:
* https
* git+https
* git+ssh

If an unsupported scheme is passed to the resolver, it can throw an error. How the url is used is also decided by the resolver.
The resolver could choose to use the url as is to handle the resolution request or it could choose to parse and extract the necessary information to construct a different url altogether when performing the resolution. This will partly be dictated by the libraries that the underlying resolver uses. 

Effectively, we want to pass it a string value. We are saying that the best practice is to format that string uniformly using the prescribed template. But how that value is used is determined by the resolver.

Reply:
It is completely ok for the resolver to be able to parse multiple formats:

e.g. 
if scheme == `https`, 
Then it must look like
```
https://<complete-git-repo>@<version/sha>#<sub path>

e.g.

https://github.com/tektoncd/pipeline@XXXXX#a/b/c
```

But if scheme == `scmType`
It must look like:
```
<scmType>://<org>/<repo>@<version>#<sub path>?<tokens etc.>

e.g.
github://tektoncd/pipeline@sha1:nsjdhjshjdh#examples/v1/taskruns/hello.yaml
```

The user can provide either format they choose. And as long as the resolver can parse the string, it's all good.

If users make a mistake while providing the string, the resolver can throw a validation error.

Reply:
May be I can make this a bit more explicit with examples and scenarios.

Reply:
Yes, I agree with all that you said I think.

Reply:
Done.